### ISSUE TITLE
opm-parser cannot process the norne and SPE9_CP decks

### ISSUE THREAD

**Issue by andlaus** [OPM/opm-parser#713]
With the latest master versions of opm-parser (revision 5c4a6385ce03d1677a03d0a4749407e80fadf318) and opm-autodiff (revision OPM/opm-autodiff@77ea648e6a528df128d8818347b70484ff62de65), neither SPE9_CP nor the Norne decks from opm-data seem to work anymore: for SPE9_CP, I get

```
---------------    Reading parameters     ---------------
output not found. Using default value 'true'.
output_dir not found. Using default value 'output'.
deck_filename found at /, value is /home/and/src/opm-data/spe9/SPE9_CP.DATA
Failed to create valid EclipseState object. See logfile: output/SPE9_CP.PRT
Exception caught: Error: expected 'double' lexemes, got: '
Program threw an exception: Error: expected 'double' lexemes, got: '
```

while Norne produces:

```
---------------    Reading parameters     ---------------
output not found. Using default value 'true'.
output_dir not found. Using default value 'output'.
deck_filename found at /, value is /home/and/src/opm-data/norne/NORNE_ATW2013.DATA
Failed to create valid EclipseState object. See logfile: output/NORNE_ATW2013.PRT
Exception caught: Error: expected all digits, got: '
Program threw an exception: Error: expected all digits, got: '
```

It is possible that this is specific to my machine, but since I have a very common distribution (ubuntu 15.10) and a somewhat common locale (German), I'm probably not alone with this...

**Comment by joakim-hove** [comment]
We are on it - seems to be some undefined behavior which works as it "should" on the Statoil tool-chain.

**Comment by andlaus** [comment]
ok, good. until this is fixed, I'll just revert my working-copy back to the latest revision of friday...

**Comment by andlaus** [comment]
BTW: are you aware of UBSAN and ASAN? ( http://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html and http://clang.llvm.org/docs/AddressSanitizer.html )

**Comment by bska** [comment]
On a possibly related note: The tests

 * `runSectionTests`
 * `runScheduleTests`

fail in a checked-iterator build of opm-parser (i.e., using libstdc++'s debug mode activated through the preprocessing symbols `-D_GLIBCXX_DEBUG=1` and `-D_GLIBCXX_DEBUG_PEDANTIC=1`).  As far as I can tell, the first failure is due to the iterator members `Opm::DeckView::first` and `Opm::DeckView::last` being default constructed--and therefore not forming a valid range--through the defaulted default constructor of `Opm::DeckView`.

The second test failure (`runScheduleTests`) is due to dereferencing null pointers in `Deck::getActiveUnitSystem()` and `Deck::getDefaultUnitSystem()`.

I'll send a copy of my `Testing/Temporary/LastTest.log` (offline) to anyone who wants it.  Note that using the debug mode changes the ABI of the STL and therefore technically requires a full rebuild of any required library component (typically Boost) that uses an iterator and/or a fully formed STL object (e.g., `std::vector<>`).

**Comment by jokva** [comment]
This has probably been fixed in @atgeirr's PR https://github.com/OPM/opm-parser/pull/714 which fixes some undefined behaviour.

**Comment by atgeirr** [comment]
I do not think that I fixed all issues @bska pointed out, and believe they must be investigated further.

**Comment by jokva** [comment]
I should probably have clarified that I referred to the original issue, not @bska's comment. I'm looking into that now.

**Comment by jokva** [comment]
I'm currently doing some investigation and figured I should report my findings so far.

Traversing over a default constructed Deck (only used in testing code afaik) was previously undefined behaviour, but this is fixed in a local patch. This affects the `runParserTests` among others. Furthermore, it seems like `boost::match_regex` *overwrites* a stack variable under some conditions, and I'm investigating why.

There is an ongoing effort to find and remove some of these typically memory errors.

**Comment by joakim-hove** [comment]
Hmmm - if there is still  boost_regex use this should be the oppurtunity to replace it with std::regex.

**Comment by andlaus** [comment]
> Hmmm - if there is still boost_regex use this should be the oppurtunity to replace it with std::regex.

I support this, but I am not the one who cares about GCC 4.8 compatibility ;)

(but boost::regex could be used unconditionally.)

**Comment by jokva** [comment]
@andlaus Could you please test with https://github.com/OPM/opm-parser/pull/718?

Furthermore, more testing indicates that `boost::match_regex` indeed writes outside of some allocated block and overwrites the memory of a stack alloc'd string. Both Valgrind and gdb reports the same error. By upgrading to gcc4.9 and `std::regex` this problem goes away, so this problem may very well be in the boost implementation.

I'm do not know if optimisation levels affect this at all.

**Comment by jokva** [comment]
I did some more testing. Consider the following program:

```
#include <string>

#include <boost/regex.hpp>

struct foo {
    foo( const std::string& ex ) : re( ex ) {} 

    bool inner_fun( const std::string& in ) const {
        return this->match( in );
    }    

    bool match( const std::string& in ) const {
        return boost::regex_match( in, this->re );
    }    

    boost::regex re;
};

bool outer_fun( const std::string& str, const foo& f ) {
    return f.inner_fun( str );
}

int main() {
    const std::string re = "BU.+|BTIPF.+|BTIPS.+|BTCNF.+|BTCNS.+|BTCN[1-9][0-9]*.+|BTIPT.+|BTIPF.+|BTIPS.+|BTIP[1-9][0-9]*.+|BTADS.+|BTDCY";                                                                                                                                      
    foo m( re );

    const std::string test = "TEST1";
    outer_fun( test, m ); 
}
```

Compiled with `g++-4.9 -Os -g -DDEBUG boost-regex.cpp -lboost_regex-mt` this runs without problems, and valgrind is happy. It's also fine with g++-4.8.

Now, if I compile it with `g++ -Os -g -D_GLIBCXX_DEBUG=1 -D_GLIBCXX_DEBUG_PEDANTIC=1 -DDEBUG boost-regex.cpp -lboost_regex-mt` it breaks the program and overwrites my data. Backtrace from when compiled with g++-4.8:

```
#11 inner_fun (
    in='\000' <repeats 24 times>, "@Ra\000\000\000\000\000\360Ta", '\000' <repeats 13 times>, "ضa\000\000\000\000\000n\000\000\000\000\000\000\000\001\000\000\000\000\000\000\000p\221a\000\000\000\000\000\000\000\000\000\004", '\000' <repeats 65 times>, "\003", '\000' <repeats 49 times>..., this=0x7fffffffd260) at boost-regex.cpp:9
#12 outer_fun (str="TEST1", f=...) at boost-regex.cpp:20
```

There's even Arabic in there. g++-4.9 behaves similarly, but spits out different data in the backtrace. Apparently, the interaction between `boost::regex` and `-D_GLIBCXX_DEBUG` triggers undefined behaviour and blows up. In other words, this should not really be an issue.

Please test PR#718 and verify that everything works as expected. It introduces a small performance regression (<10ms, but still) but should hopefully fix the errors.

And thanks for the report.

**Comment by bska** [comment]
> if I compile it with
```sh
g++ -Os -g -D_GLIBCXX_DEBUG=1 -D_GLIBCXX_DEBUG_PEDANTIC=1 -DDEBUG boost-regex.cpp -lboost_regex-mt
```

then you also need a Boost library that's build with those debug options.  Otherwise you'll trigger false positives.

**Comment by andlaus** [comment]
> Furthermore, more testing indicates that boost::match_regex indeed writes outside of some allocated block and overwrites the memory of a stack alloc'd string. Both Valgrind and gdb reports the same error. By upgrading to gcc4.9 and std::regex this problem goes away, so this problem may very well be in the boost implementation.

this could be, but then it is a problem with recent compilers because I'm pretty sure that I ran flow for norne with valgrind some time ago and it did not report anything.

the question is: *if* boost::regex does not work anymore, how can GCC 4.8 be supported (std::regex is only implemented properly for GCC >= 4.9)

**Comment by jokva** [comment]
> then you also need a Boost library that's built with those debug options. Otherwise you'll trigger false positives.

Which probably was what happened, and the reason for all these shenanigans.

>  the question is: if boost::regex does not work anymore, how can GCC 4.8 be supported (std::regex is only implemented properly for GCC >= 4.9)

The errors showed up with the debug flags (the gcc regex was obviously compiled with them as they were compiled and instantiated with the same configuration), so probably binary incompatibility between the boost libs and my local version. This probably isn't an issue.
